// @teppeis/multimaps/dist/esm/arraymultimap.js
/*!
 * melonJS Game Engine - v15.15.0
 * http://www.melonjs.org
 * melonjs is licensed under the MIT License.
 * http://www.opensource.org/licenses/mit-license
 * @copyright (C) 2011 - 2023 Olivier Biot (AltByte Pte Ltd)
 */
import { Multimap } from './multimap.js';

class ArrayMultimap extends Multimap {
    constructor(iterable) {
        super(new ArrayOperator(), iterable);
    }
    get [Symbol.toStringTag]() {
        return "ArrayMultimap";
    }
}
class ArrayOperator {
    create() {
        return [];
    }
    clone(collection) {
        return collection.slice();
    }
    add(value, collection) {
        collection.push(value);
        return true;
    }
    size(collection) {
        return collection.length;
    }
    delete(value, collection) {
        const index = collection.indexOf(value);
        if (index > -1) {
            collection.splice(index, 1);
            return true;
        }
        return false;
    }
    has(value, collection) {
        return collection.includes(value);
    }
}

export { ArrayMultimap };
